### DISCUSSION TITLE
Icecast Protocol Implementation Help

### DISCUSSION THREAD

**Discussion by aw-was-here** [aio-libs/aiohttp#7086] [Q&A] [answered]
Hello!

I'm trying to implement an Icecast server using aiohttp but I'm quite stuck. Let me describe the protocol and where I'm hung up.

Protocol is fairly trivial:

- Client connects with the SOURCE method.
- Server responds with 200.
- On the same socket, client now starts sending binary data.

I've got a route set up to accept the SOURCE method to my handler which then does something relatively trivial:

```python
async def icecast_source_handler(self, request):
    response = web.StreamResponse(status=200)
    await response.prepare(request)
    reader = request.content
  ... rest of code here ...
```

but reader is basically not getting new data and ends up being an EmptyStreamReader (I'm guessing because there is no body to read?)  How do I hang onto the socket to process more incoming data?

FWIW, here is an [implementation using http.server](https://github.com/radusuciu/traktor_nowplaying/blob/de8be51372c90c523b49ca8052c817c9c687dc91/traktor_nowplaying/core.py#L21).

Thanks!

**Comment by Dreamsorcerer** [comment]
A complete client/server example that we can run would help. I'd expect that if the client connection is still open and no EOF has been sent, then `await request.content.read()` should work...

Although it sounds like the Icecast protocol should really be using the Expect header, which aiohttp already has extra code to handle (what you describe is basically an expect handler; client sends `Expect: 100-continue` to server before trying to send a large payload, server responds with `100 Continue`, client then proceeds to send the payload):
https://docs.aiohttp.org/en/stable/web_advanced.html#expect-header

**Reply by aw-was-here** [reply]
Thank you for the response.

Unfortunately the Shoutcast/Icecast protocols are from 1998 and pre-date a lot of the HTTP 1.1 features like Expect.  A newer spec came out in the past few years that use PUT+Expect but the vast majority of clients do not support it yet.

I pointed to a simple server implementation that doesn't use async above.  Let me see if I can locate a Python client.  In the meantime, the ices binary, libshout, Mixxx, Traktor, and others all implement the client side.

**Reply by Dreamsorcerer** [reply]
> I pointed to a simple server implementation that doesn't use async above. Let me see if I can locate a Python client. In the meantime, the ices binary, libshout, Mixxx, Traktor, and others all implement the client side.

Yes, I mean the aiohttp one. Whatever code you have to reproduce the problem, plus some client code to trigger it. We're rather starved for time, so anything that requires more than a minute to figure out how to reproduce is unlikely to get one of us to look at. ;)

**Comment by aw-was-here** [comment, marked as answer]
OK, I've pretty much given up on using aiohttp for this project and instead went for a custom [asycnio.Protocol](https://github.com/whatsnowplaying/whats-now-playing/blob/main/nowplaying/inputs/icecast.py#L38).  It isn't very smart handling http but it seems to be working well enough for Traktor, [MIXX](https://github.com/mixxxdj/mixxx), and [butt](https://danielnoethen.de/butt/) to use.

Thank you for your time.
